modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/TcpCommunicationMetricsListener.java

```diff
@@ -17,74 +17,162 @@
  */
-public class TcpCommunicationMetricsListener implements GridNioMetricsListener{
-    /** Counter factory. */
-    private static final Callable<LongHolder> HOLDER_FACTORY = new Callable<LongHolder>() {
-        @Override public LongHolder call() {
-            return new LongHolder();
-        }
-    };
+class TcpCommunicationMetricsListener {
+    /** Metrics manager. */
+    private final GridMetricManager mmgr;
 
-    /** Received bytes count. */
-    private final LongAdder rcvdBytesCnt = new LongAdder();
+    /** Metrics registry. */
+    private final MetricRegistry mreg;
 
-    /** Sent bytes count.*/
-    private final LongAdder sentBytesCnt = new LongAdder();
+    /** */
+    private final Function<Short, AtomicLongMetric> sentMsgsCntByTypeMetricFactory;
 
-    /** All registered metrics. */
-    private final Set<ThreadMetrics> allMetrics = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    /** */
+    private final Function<Short, AtomicLongMetric> rcvdMsgsCntByTypeMetricFactory;
 
-    /** Thread-local metrics. */
-    private final ThreadLocal<ThreadMetrics> threadMetrics = new ThreadLocal<ThreadMetrics>() {
-        @Override protected ThreadMetrics initialValue() {
-            ThreadMetrics metrics = new ThreadMetrics();
+    /** */
+    private final Function<Object, AtomicLongMetric> sentMsgsCntByConsistentIdMetricFactory;
 
-            allMetrics.add(metrics);
+    /** */
+    private final Function<Object, AtomicLongMetric> rcvdMsgsCntByConsistentIdMetricFactory;
 
-            return metrics;
-        }
-    };
+    /** Sent bytes count metric.*/
+    private final AtomicLongMetric sentBytesMetric;
+
+    /** Received bytes count metric. */
+    private final AtomicLongMetric rcvdBytesMetric;
+
+    /** Sent messages count metric. */
+    private final AtomicLongMetric sentMsgsMetric;
+
+    /** Received messages count metric. */
+    private final AtomicLongMetric rcvdMsgsMetric;
+
+    /** Sent messages count metrics grouped by message type. */
+    ConcurrentHashMap<Short, AtomicLongMetric> sentMsgsMetricsByType = new ConcurrentHashMap<>();
+
+    /** Received messages count metrics grouped by message type. */
+    ConcurrentHashMap<Short, AtomicLongMetric> rcvdMsgsMetricsByType = new ConcurrentHashMap<>();
+
+    /**
+     * Sent messages count metrics grouped by message node id.
+     * @deprecated for JMX support only.
+     */
+    @Deprecated
+    ConcurrentHashMap<UUID, AtomicLong> sentMsgsMetricsByNodeId = new ConcurrentHashMap<>();
+
+    /**
+     * Received messages metrics count grouped by message node id.
+     * @deprecated for JMX support only.
+     */
+    @Deprecated
+    ConcurrentHashMap<UUID, AtomicLong> rcvdMsgsMetricsByNodeId = new ConcurrentHashMap<>();
+
+    /** Sent messages count metrics grouped by message node consistent id. */
+    ConcurrentHashMap<Object, AtomicLongMetric> sentMsgsMetricsByConsistentId = new ConcurrentHashMap<>();
```

Comment:
I'm not sure we should change the way we collect metrics in this PR.
Let's keep existing behaviour?

This also will simplify PR a lot.

Reply:
I changed that because in old implementation we stored basically the same map in each thread separately. That's the obvious answer.

The other thing is that old implementation had concurrency issues that no one cared about. Like broken values visibility щк potential ConcurrentModificationExceptions. I wanted new implementation to be correct.

Reply:
You introduced metrics based on node consistency id.
I think we should avoid introducing new metrics here.

The only goal of this ticket is to migrate *existing* metrics to the new framework.

Let's 
1. remove metrics based on consistency id.
2. revert changes in `ConnectionKey` and `TcpCommunicationSpi` that made to support it.